### PR TITLE
BitConverter tests: widen NaN check

### PR DESCRIPTION
**Pull request by lewurm**
Before we were assuming `float.NaN` is encoded as `0xffc00000`,
however, this is only one of many valid representations for `NaN`.

According to [1], NaN is encoded as either:
* signalling NaN (NANS) is represented by any bit pattern between 0x7f800001
  and 0x7fbfffff or between 0xff800001 and 0xffbfffff
* or, quiet NaN (NANQ) is represented by any bit pattern between 0x7fc00000 and
  0x7fffffff or between 0xffc00000 and 0xffffffff

We only care about quiet NaN in this case.

Likewise for `double`.

[1] https://www.doc.ic.ac.uk/~eedwards/compsys/float/nan.html

Fixes https://github.com/mono/mono/issues/10963

### CODE REVIEW

**Comment by stephentoub** [comment]
cc: @tannergooding

**Comment by jkotas** [comment]
`float.NaN` and `double.NaN` are constants. I do not think we want to allow ambiguity on whether it is quiet NaN or signaling NaN.

The equivalent C constant http://www.cplusplus.com/reference/cmath/NAN/ is explicitly specified as quiet NaN. They do not allow ambiquity on whether it is one or the other either.

**Comment by lewurm** [comment]
> `float.NaN` and `double.NaN` are constants.

https://github.com/dotnet/corefx/blob/e7af146143c26d547b565a671713d72146047eb8/src/Common/src/CoreLib/System/Single.cs#L38
They are defined as an expression. Maybe that's the culprit?

> I do not think we want to allow ambiguity on whether it is quiet NaN or signaling NaN.

In that case it's really about that the most significant bit is not relevant; for example armv7 produces a different result than amd64 for the expression above differ by the most significant bit.

**Comment by jkotas** [comment]
> They are defined as an expression. Maybe that's the culprit?

`0.0/0.0` expression is the only way to define NaN in C# today. C# compiler should be deterministic, independent on the platform. `0.0/0.0` should compile to same bits everywhere. The best place to fix may be to add normalization of NaNs to Roslyn. @jaredpar Do you have an opinion?

**Comment by tannergooding** [comment]
> I do not think we want to allow ambiguity on whether it is quiet NaN or signaling NaN.

Their is no ambiguity in the result, it is quiet with no payload for both 0x7FC and 0xFFC. The only difference is whether the sign bit is set.

> C# compiler should be deterministic, independent on the platform. 0.0/0.0 should compile to same bits everywhere.

I agree it would be nice for the compiler to be deterministic here (and IEEE allows it under the list of "value changing transformations which preserve the literal meaning of the source code" -- see IEEE 754:2008 10.4). However, the current difference is catching bugs in our and Mono's code (even if it is currently test code), and changing this may mask other bugs that exist.

File: src/System.Runtime.Extensions/tests/System/BitConverter.cs [hunk 1]
```diff
@@ -115,6 +115,22 @@ public static void DoubleToInt64Bits()
             Assert.Equal(input, roundtripped);
         }
 
+		[Fact]
+		public static void SingleNaNToInt32Bits()
+		{
+			int bits = BitConverter.SingleToInt32Bits(float.NaN);
+			Assert.Equal (0x7f800000, bits & 0x7f800000);
```
Comment (tannergooding) [inline src/System.Runtime.Extensions/tests/System/BitConverter.cs]: Why not just assert that `(bits & 0x7FFFFFFF) > 0x7FC00000` (which is basically `IsQuietNaN`)?

**Comment by jaredpar** [comment]
> The best place to fix may be to add normalization of NaNs to Roslyn. @jaredpar Do you have an opinion?

That makes sense to me. @gafter do you agree?

File: src/System.Runtime.Extensions/tests/System/BitConverter.cs (hunk 1 above)
Comment (lewurm) [reply src/System.Runtime.Extensions/tests/System/BitConverter.cs]: I didn't want to copy/paste it 😄 but that isn't a good reason, I'll change it, as it's easier to read.

Any reason why you suggest to restrict it to `0x7fc00000` (which will catch quiet NaNs only)? `0x7f800000` would catch all representations of NaN. Is it expected that `float.NaN` is _always_ a quiet NaN?

Comment (tannergooding) [reply src/System.Runtime.Extensions/tests/System/BitConverter.cs]: > Is it expected that float.NaN is always a quiet NaN?

Yes.

Comment (lewurm) [reply src/System.Runtime.Extensions/tests/System/BitConverter.cs]: Thanks, updated.

Comment (lewurm) [reply src/System.Runtime.Extensions/tests/System/BitConverter.cs]: Tiny correction: It must be `(bits & 0x7FFFFFFF) >= 0x7FC00000`.

Comment (jkotas) [reply src/System.Runtime.Extensions/tests/System/BitConverter.cs]: This relaxation of the check would make sense for value computed at runtime. For example, if we had:

```
static double x = 0.0;

static void Test()
{
    double y = x/x;
    ... it is fine to assert that y matches any valid NaN pattern ...
}
```

However, this relaxation does not make sense to me for validation of a NaN constant. It is working around Roslyn determinism bug - Roslyn is expected to produce a deterministic output.

**Comment by gafter** [comment]
I agree. I'd say our failure to normalize NaNs is a bug in Roslyn's determinism.

**Comment by gafter** [comment]
Wait... the CLR should be producing exactly the same bits for arithmetic like `0.0 / 0.0` on every platform. Roslyn needs to depend on the runtime to produce correct results in order to do its job. If runtimes vary, I'd think that a runtime bug. But perhaps the CLI specification doesn't nail down the value required for `0.0 / 0.0`, in which case it may be a CLI spec bug.

Roslyn can normalize the result if there are bugs in some implementations to work around those bugs, but only when we become aware of them. We are now aware of this, so I've added https://github.com/dotnet/roslyn/issues/30439 to address it.

**Comment by jkotas** [comment]
> Wait... the CLR should be producing exactly the same bits for arithmetic like `0.0` / `0.0` on every platform. 

@gafter IEEE (that the CLR follows) does not prescribe the exact NaN bit pattern. Some of the NaN bits can have non-deterministic values.

**Comment by jkotas** [comment]
It should be only the latter. Thanks for looking into this.

**Pull request closed without merging**